File: r-app/src-tauri/src/models/usage.rs
use serde::{Deserialize, Serialize};

/// 内部：单条本机用量记录（落 `usage_records`）。
/// 语义统一为「输入=非缓存提示 token，缓存读取/创建单列」，便于跨 app 汇总不重复计数。
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct UsageRecord {
    pub app_type: String,   // "claude" | "codex"
    pub record_key: String, // 去重键（claude: message.id；codex: codex_session:<id>:<idx>）
    pub date: String,       // YYYY-MM-DD（本地时区）
    pub model: String,
    pub requests: i64,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cache_creation_tokens: i64,
    pub cache_read_tokens: i64,
}

/// 用量总览（MVP：仅 token 维度，无成本）。
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UsageSummary {
    pub total_requests: i64,
    pub total_input_tokens: i64,
    pub total_output_tokens: i64,
    pub total_cache_creation_tokens: i64,
    pub total_cache_read_tokens: i64,
}

/// 按模型聚合。
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelUsage {
    pub app_type: String,
    pub model: String,
    pub requests: i64,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cache_creation_tokens: i64,
    pub cache_read_tokens: i64,
}

/// 按天聚合。
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DailyUsage {
    pub date: String,
    pub app_type: String,
    pub requests: i64,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cache_creation_tokens: i64,
    pub cache_read_tokens: i64,
}

/// 按天 × 来源 × 模型聚合（多维合并表：前端按 date 行合并展示）。
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DayModelUsage {
    pub date: String,
    pub app_type: String,
    pub model: String,
    pub requests: i64,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cache_creation_tokens: i64,
    pub cache_read_tokens: i64,
}

/// 一次本机用量同步结果。
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UsageSyncResult {
    pub imported: i64,
    pub files_scanned: i64,
    pub errors: i64,
}

File: src/services/modules/usage.ts
import { request } from '../request'

export type UsageAppFilter = 'all' | 'claude' | 'codex'

export interface UsageSummary {
  totalRequests: number
  totalInputTokens: number
  totalOutputTokens: number
  totalCacheCreationTokens: number
  totalCacheReadTokens: number
}

/** 按天 × 来源 × 模型聚合（多维合并表：前端按 date 行合并展示）。 */
export interface DayModelUsage {
  date: string
  appType: string
  model: string
  requests: number
  inputTokens: number
  outputTokens: number
  cacheCreationTokens: number
  cacheReadTokens: number
}

export interface UsageSyncResult {
  imported: number
  filesScanned: number
  errors: number
}

interface UsageFilter {
  start?: string
  end?: string
  appType?: string
}

export const usageApi = {
  /** 触发本机用量增量同步。 */
  sync: () => request<UsageSyncResult>('sync_session_usage'),
  getSummary: (f: UsageFilter = {}) =>
    request<UsageSummary>('get_usage_summary', {
      start: f.start,
      end: f.end,
      appType: f.appType,
    }),
  /** 按天 × 来源 × 模型聚合（date 倒序、组内 token 降序）。 */
  getByDayModel: (f: UsageFilter = {}) =>
    request<DayModelUsage[]>('get_usage_by_day_model', {
      start: f.start,
      end: f.end,
      appType: f.appType,
    }),
}
